com.unity.shadergraph/Editor/Data/Nodes/Math/Advanced/RefractNode.cs

```diff
@@ -0,0 +1,80 @@
+using System.Reflection;
+using UnityEngine;
+using UnityEditor.Graphing;
+using UnityEditor.ShaderGraph.Drawing.Controls;
+
+namespace UnityEditor.ShaderGraph
+{
+    enum RefractMode
+    {
+        CriticalAngle,
+        Safe
+    };
+
+    [Title("Math", "Advanced", "Refract")]
+    class RefractNode : CodeFunctionNode
+    {
+        [SerializeField]
+        private RefractMode m_RefractMode = RefractMode.Safe;
+
+        [EnumControl("Mode")]
+        public RefractMode refractMode
+        {
+            get { return m_RefractMode; }
+            set
+            {
+                if (m_RefractMode == value)
+                    return;
+
+                m_RefractMode = value;
+                Dirty(ModificationScope.Graph);
+            }
+        }
+
+        public RefractNode()
+        {
+            name = "Refract";
+            synonyms = new string[] { "refract", "refract safe", "refract critical angle" };
+        }
+
+        protected override MethodInfo GetFunctionToConvert()
+        {
+            if (m_RefractMode == RefractMode.CriticalAngle)
+                return GetType().GetMethod("Unity_RefractCriticalAngle", BindingFlags.Static | BindingFlags.NonPublic);
+            else
+                return GetType().GetMethod("Unity_RefractSafe", BindingFlags.Static | BindingFlags.NonPublic);
+        }
+
+        static string Unity_RefractCriticalAngle(
+            [Slot(0, Binding.None, 0, 0, 0, 1)] DynamicDimensionVector Incident,
+            [Slot(1, Binding.None, 0, 1, 0, 1)] DynamicDimensionVector Normal,
+            [Slot(2, Binding.None)] Vector1 Eta,
+            [Slot(3, Binding.None)] out DynamicDimensionVector Out)
+        {
+            return
+@"
+{
```

Comment:
Hmm, I expected you would use the refract() intrinsic function. How is this different from that?

Reply:
"intrinsic" refract will return zero at the critical angle and our case we have 2 modes criticalAngle and safe to handle that.
For the safe-mode we block the rectracted angle to the surface, and for the critical here we return let the math do the rest.